### PR TITLE
[MPS] Add nondeterministic alerts for ops that use `atomic_add`

### LINKED ISSUE #195404: [MPS] `Embedding.backward()` is non-deterministic
### 🐛 Describe the bug

On the MPS back end, `Embedding`'s `backward()` can be non-deterministic, even when `torch.use_deterministic_algorithms(True)` is used, as demonstrated below:

```python
import torch
import sys

C = 8
D = 16
B = 10
T = 32

def run_embedding(weights, t, m):
  embedding = torch.nn.Embedding(C, D).to("mps")
  embedding.load_state_dict(weights)
  loss = (embedding(t) * m).sum()
  loss.backward()
  torch.mps.synchronize()
  return embedding.weight.grad

def run_manual(weights, t, m):
  embedding = torch.nn.Embedding(C, D).to("mps")
  embedding.load_state_dict(weights)
  one_hot = torch.nn.functional.one_hot(t, num_classes=C).to(embedding.weight.dtype)
  loss = ((one_hot @ embedding.weight) * m).sum()
  loss.backward()
  torch.mps.synchronize()
  return embedding.weight.grad

def run(fn, name, weights, t, m):
  trials = 20
  vals = [fn(weights, t, m) for _ in range(trials+1)]
  equals = [torch.equal(vals[i], vals[i+1]) for i in range(trials)]
  n_equal = sum(equals)
  print(f"{name} has {n_equal}/{trials} equal gradients")

if not torch.backends.mps.is_available():
  sys.exit(1)
torch.use_deterministic_algorithms(True)
torch.manual_seed(0)
weights = torch.nn.Embedding(C, D).state_dict()
t = torch.arange(B * T, device="mps").reshape(B, T) % C
m = torch.arange(B * T * D, dtype=torch.float32, device="mps").reshape(B, T, D) / 1000

run(run_embedding, "run_embedding", weights, t, m)
run(run_manual, "run_manual", weights, t, m)

a = run_embedding(weights, t, m)
b = run_manual(weights, t, m)
print(torch.allclose(a, b))
print(torch.__version__)
```

Output:
```
run_embedding has 0/20 equal gradients
run_manual has 20/20 equal gradients
True
2.13.0
```

It's [known](https://github.com/pytorch/pytorch/blob/v2.13.0/aten/src/ATen/native/mps/operations/Indexing.mm#L200) that FP atomic add is non-associative and thus can race. A low number of embeddings may contribute due to increased possibility of addition "collisions" when accumulating the gradients.

I was encouraged to file this after reading the [Reproducibility](https://docs.pytorch.org/docs/2.13/notes/randomness.html) documentation. Ideally, either a deterministic alternative is enabled by `torch.use_deterministic_algorithms(True)`, or an error message is issued.

A search did not yield directly relevant prior issues. #97236 discusses non-specific MPS backend non-determinism, although the code presented in that issue does appear to involve `nn.Embedding`.

### Versions

Collecting environment information...
PyTorch version: 2.13.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.5 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.0.123.102)
CMake version: Could not collect
Libc version: N/A

Python version: 3.14.6 | packaged by conda-forge | (main, Jul 24 2026, 16:05:26) [Clang 20.1.8 ] (64-bit runtime)
Python platform: macOS-26.5-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] Could not collect
[conda] libtorch                     2.13.0           cpu_generic_hf893bd3_2        conda-forge
[conda] nomkl                        1.0              h5ca1d4c_0                    conda-forge
[conda] numpy                        2.5.1            py314hb79c6fa_0               conda-forge
[conda] optree                       0.19.1           py314h6cfcd04_0               conda-forge
[conda] pytorch                      2.13.0           cpu_generic_py314_hb7ad2ff_2  conda-forge

cc @mruberry @kurtamohler @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by kurtamohler**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #195754

Also updates `use_deterministic_algorithms` documentation to describe
all cases when MPS ops raise an error, and adds a test for them.

Fixes #195404

Co-authored by Claude

### CODE REVIEW

**Review by malfet — APPROVED** [review]

File: torch/__init__.py [hunk 1]
```diff
@@ -1837,6 +1837,28 @@ def use_deterministic_algorithms(
         * :func:`torch.median` with indices output when called on a CUDA tensor
         * :func:`torch.nn.functional.grid_sample` when attempting to differentiate a CUDA tensor
         * :func:`torch.Tensor.scatter_reduce` when ``reduce='prod'`` and called on CUDA tensor
```
Comment (malfet) [inline torch/__init__.py]: Why not extend it to MPS as well? Also, are you sure that sum is deterministic on CUDA?

File: torch/__init__.py [hunk 2]
```diff
@@ -1837,6 +1837,28 @@ def use_deterministic_algorithms(
...
+          an MPS tensor with floating or complex dtype
+        * :func:`torch.Tensor.index_reduce` with ``reduce='prod'\'mean'``
+          when called on an MPS tensor with floating or complex dtype
+        * :func:`torch.kthvalue` when called on an MPS tensor
+        * :class:`torch.nn.Embedding` when attempting to differentiate an MPS tensor
+        * :class:`torch.nn.EmbeddingBag` when attempting to differentiate an MPS tensor
+        * :func:`torch.nn.functional.grid_sample` when attempting to differentiate an MPS tensor
+        * :func:`torch.nn.MaxPool2d` when attempting to differentiate an MPS tensor
+        * :func:`torch.nn.MaxPool3d` when attempting to differentiate an MPS tensor
+        * :func:`torch.nn.AvgPool3d` when attempting to differentiate an MPS tensor
```
Comment (malfet) [inline torch/__init__.py]: Should you express this one on one line? I.e. MaxPool[23]d and AvgPool3d

File: torch/__init__.py [hunk 3]
```diff
@@ -1837,6 +1837,28 @@ def use_deterministic_algorithms(
...
+        * :func:`torch.nn.functional.interpolate` when attempting to differentiate an MPS tensor
+          in one of the following cases:
+
+          - ``mode='bilinear'`` and ``antialias=True``
+          - ``mode='bicubic'``
+          - ``mode='trilinear'``
+          - ``mode='nearest'`` with 5-D input
+          - ``mode='nearest-exact'`` with 5-D input
```
Comment (malfet) [inline torch/__init__.py]: I'm not sure if excessive documentation is necessary here, just vague in multiple classes is probably fine

File: torch/__init__.py (hunk 1 above)
Comment (kurtamohler) [reply torch/__init__.py]: Good catch, the alert is raised in either case for CUDA: https://github.com/pytorch/pytorch/blob/f3238ae32715ec5fa7df5574907b6bd6d4b59306/aten/src/ATen/native/cuda/ScatterGatherKernel.cu#L589
